middleware/cors/utils.go

```diff
@@ -1,56 +1,85 @@
+	normalizedPattern := normalizeDomain(pattern)
+
+	// Handling the case where pattern is a wildcard subdomain pattern.
+	if strings.HasPrefix(normalizedPattern, "*.") {
+		// Trim leading "*." from pattern for comparison.
+		trimmedPattern := normalizedPattern[2:]
+
+		// Check if the domain ends with the trimmed pattern.
+		if strings.HasSuffix(normalizedDomain, trimmedPattern) {
+			// Ensure that the domain is not exactly the base domain.
+			if normalizedDomain != trimmedPattern {
+				// Special handling to prevent "example.com" matching "*.example.com".
+				if strings.TrimSuffix(normalizedDomain, trimmedPattern) != "" {
+					return true
+				}
+			}
+		}
 	}
-	domAuth := domain[didx+3:]
-	// to avoid long loop by invalid long domain
-	const maxDomainLen = 253
-	if len(domAuth) > maxDomainLen {
-		return false
+
+	return false
+}
+
+// normalizeDomain removes the scheme and port from the input domain
+func normalizeDomain(input string) string {
+	// Remove scheme
+	input = strings.TrimPrefix(strings.TrimPrefix(input, "http://"), "https://")
+
+	// Find and remove port, if present
+	if len(input) > 0 && input[0] != '[' {
+		if portIndex := strings.Index(input, ":"); portIndex != -1 {
+			input = input[:portIndex]
+		}
 	}
-	patAuth := pattern[pidx+3:]
-
-	domComp := strings.Split(domAuth, ".")
-	patComp := strings.Split(patAuth, ".")
-	const divHalf = 2
-	for i := len(domComp)/divHalf - 1; i >= 0; i-- {
-		opp := len(domComp) - 1 - i
-		domComp[i], domComp[opp] = domComp[opp], domComp[i]
+
+	return input
+}
+
+// normalizeOrigin checks if the provided origin is in a correct format
+// and normalizes it by removing any path or trailing slash.
+// It returns a boolean indicating whether the origin is valid
+// and the normalized origin.
+func normalizeOrigin(origin string) (bool, string) {
+	parsedOrigin, err := url.Parse(origin)
+	if err != nil {
+		return false, ""
 	}
-	for i := len(patComp)/divHalf - 1; i >= 0; i-- {
-		opp := len(patComp) - 1 - i
-		patComp[i], patComp[opp] = patComp[opp], patComp[i]
+
+	// Validate the scheme is either http or https
+	if parsedOrigin.Scheme != "http" && parsedOrigin.Scheme != "https" {
+		return false, ""
 	}
 
-	for i, v := range domComp {
-		if len(patComp) <= i {
-			return false
-		}
-		p := patComp[i]
-		if p == "*" {
-			return true
-		}
-		if p != v {
-			return false
-		}
+	// Validate there is a host present. The presence of a path, query, or fragment components
+	// is checked, but a trailing "/" (indicative of the root) is allowed for the path and will be normalized
+	if parsedOrigin.Host == "" || (parsedOrigin.Path != "" && parsedOrigin.Path != "/") || parsedOrigin.RawQuery != "" || parsedOrigin.Fragment != "" {
+		return false, ""
 	}
-	return false
+
+	// Normalize the origin by constructing it from the scheme and host.
+	// The path or trailing slash is not included in the normalized origin.
+	return true, strings.ToLower(parsedOrigin.Scheme) + "://" + strings.ToLower(parsedOrigin.Host)
```

Comment:
Here we can reduce processing from the CPU.
```suggestion
	return true, strings.ToLower(parsedOrigin.Scheme + "://" + parsedOrigin.Host)
```

benchmark:
```
// BenchmarkBothLower-8   	16690329	        62.00 ns/op	      32 B/op	       1 allocs/op
// BenchmarkOneLower-8   	22892752	        50.97 ns/op	       0 B/op	       0 allocs/op
```

code
```go
func bothLower(scheme, domain string) string {
	return strings.ToLower(scheme) + "://" + strings.ToLower(domain)
}

func oneLower(scheme, domain string) string {
	return strings.ToLower(scheme + "://" + domain)
}

func BenchmarkBothLower(b *testing.B) {
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = bothLower("https", "renanbastos.com.br")
	}
}

func BenchmarkOneLower(b *testing.B) {
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = oneLower("https", "renanbastos.com.br")
	}
}
```